cmd/kubeadm/app/cmd/upgrade/apply.go

```diff
@@ -0,0 +1,192 @@
+		return err
+	}
+
+	// If interactive mode is set (--yes was not specified), ask ther user whether they really want to upgrade
+	// If --yes is specified; just proceed and do it
+	if !flags.nonInteractiveMode {
+		if err := upgrade.InteractivelyConfirmUpgrade("Are you sure you want to proceed with the upgrade"); err != nil {
+			return err
+		}
+	}
+
+	// Use a prepuller implementation based on creating DaemonSets
+	// and block until all DaemonSets are ready; then we know for sure that all control plane images are cached locally
+	prepuller := upgrade.NewDaemonSetPrepuller(client, internalcfg)
+	upgrade.PrepullImagesInParallel(prepuller, flags.pullTimeout)
+
+	// Now; perform the upgrade procedure
+	if err := PerformUpgrade(flags, client, internalcfg); err != nil {
+		return err
+	}
+
+	// Upgrade RBAC rules and addons. Optionally, if needed, perform some extra task for a specific version
+	if err := upgrade.PerformPostUpgradeTasks(client, internalcfg, flags.newK8sVersion); err != nil {
+		return err
+	}
+
+	fmt.Println("")
+	fmt.Printf("[upgrade/successful] SUCCESS! Your cluster was upgraded to %q. Enjoy!\n", flags.newK8sVersionStr)
+	fmt.Println("")
+	fmt.Println("[upgrade/kubelet] Now that your control plane is upgraded, please proceed with upgrading your kubelets in turn.")
+
+	return nil
+}
+
+// ValidateFlags makes sure the arguments given to apply are valid
+func ValidateFlags(flags *applyFlags) error {
+	if len(flags.newK8sVersionStr) == 0 {
+		return fmt.Errorf("--version is a required parameter. Please specify the version you want to upgrade to")
+	}
+	var err error
+	flags.newK8sVersion, err = version.ParseSemantic(flags.newK8sVersionStr)
+	if err != nil {
+		return fmt.Errorf("couldn't parse version %q as a semantic version", flags.newK8sVersionStr)
+	}
+	return nil
+}
+
+// SetFlagDefaults handles dynamically defaulting flags based on each other's value
+func SetFlagDefaults(flags *applyFlags) {
+	// If we are in dry-run or force mode; we should automatically execute this command non-interactively
+	if flags.dryRun || flags.force {
+		flags.nonInteractiveMode = true
+	}
+}
+
+// EnforceVersionPolicies makes sure that the version the user specified is valid to upgrade to
+// There are both fatal and skippable (with --force) errors
+func EnforceVersionPolicies(flags *applyFlags, client clientset.Interface) error {
+	fmt.Printf("[upgrade/version] You have chosen to upgrade to version %q\n", flags.newK8sVersionStr)
+	versionGetterImpl := upgrade.NewKubeVersionGetter(client, os.Stdout)
+	versionSkewErrs := upgrade.EnforceVersionPolicies(versionGetterImpl, flags.newK8sVersionStr, flags.newK8sVersion, flags.parent.allowExperimentalUpgrades, flags.parent.allowRCUpgrades)
+	if versionSkewErrs != nil {
+
+		if len(versionSkewErrs.Mandatory) > 0 {
+			fmt.Printf("[upgrade/version] FATAL: The --version argument is invalid due to these fatal errors: %v.\n", versionSkewErrs.Mandatory)
```

Comment:
It seems like an antipattern to both print an error *and* return it. Shouldn't we just return a well-formatted error so that whoever chooses to handle it can print the appropriate message? Otherwise, we're asking for duplicates.